spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -2238,11 +2239,14 @@ protected void doInTransactionWithoutResult(TransactionStatus status) {
 
 		private List<ConsumerRecord<K, V>> createRecordList(final ConsumerRecords<K, V> records) {
 			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
-			List<ConsumerRecord<K, V>> list = new LinkedList<>();
+			@SuppressWarnings("unchecked") ConsumerRecord<K, V>[] recordsArray =
+					(ConsumerRecord<K, V>[]) Array.newInstance(ConsumerRecord.class, records.count());
+			int index = 0;
 			while (iterator.hasNext()) {
-				list.add(iterator.next());
+				recordsArray[index] = iterator.next();
+				index += 1;
```

Comment:
```suggestion
				recordsArray[index++] = iterator.next();
```

Reply:
Thanks.
You can apply this change yourself from this UI.